catalog/dags/data_refresh/distributed_reindex.py

```diff
@@ -0,0 +1,425 @@
+    it will run with the same version of the indexer worker as the others, even if code has been
+    deployed to the indexer worker in the meantime.
+    """
+    if environment != PRODUCTION:
+        raise AirflowSkipException("Skipping instance creation in local environment.")
+
+    ec2_hook = EC2Hook(aws_conn_id=aws_conn_id, api_type="client_type")
+    launch_templates = ec2_hook.conn.describe_launch_templates(
+        LaunchTemplateNames=INDEXER_LAUNCH_TEMPLATES.get(target_environment)
+    )
+
+    if len(launch_templates.get("LaunchTemplates")) == 0:
+        raise Exception("Unable to determine launch template version.")
+
+    return launch_templates.get("LaunchTemplates")[0].get("LatestVersionNumber")
+
+
+@task
+def create_worker(
+    environment: str,
+    target_environment: Environment,
+    launch_template_version: int,
+    aws_conn_id: str = AWS_CONN_ID,
+):
+    """
+    Create a new EC2 instance using the launch template for the target
+    environment. In local development, this step is skipped.
+    """
+    if environment != PRODUCTION:
+        raise AirflowSkipException("Skipping instance creation in local environment.")
+
+    ec2_hook = EC2Hook(aws_conn_id=aws_conn_id, api_type="client_type")
+    instances = ec2_hook.conn.run_instances(
+        MinCount=1,
+        MaxCount=1,
+        LaunchTemplate={
+            "LaunchTemplateName": INDEXER_LAUNCH_TEMPLATES.get(target_environment),
+            "Version": str(launch_template_version),
+        },
+        # Name the instance by applying a tag
+        TagSpecifications=[
+            {
+                "ResourceType": "instance",
+                "Tags": [
+                    {
+                        "Key": "Name",
+                        "Value": f"catalog-indexer-worker-{target_environment}",
+                    },
+                ],
+            },
+        ],
+    )["Instances"]
+
+    if not len(instances) == 1:
+        raise Exception(
+            f"Expected one new instance, but {len(instances)} were created."
+        )
+
+    return instances[0]["InstanceId"]
+
+
+@task.sensor(poke_interval=60, timeout=3600, mode="reschedule")
+def wait_for_worker(
+    environment: str,
+    instance_id: str,
+    aws_conn_id: str = AWS_CONN_ID,
+):
+    """Await the EC2 instance with the given id to be in a healthy running state."""
+    if environment != PRODUCTION:
+        raise AirflowSkipException("Skipping instance creation in local environment.")
+
+    ec2_hook = EC2Hook(aws_conn_id=aws_conn_id, api_type="client_type")
+    result = ec2_hook.conn.describe_instance_status(InstanceIds=[instance_id])
+
+    instance_status = result.get("InstanceStatuses", [])[0]
+    state = instance_status.get("InstanceState", {}).get("Name")
+    status = next(
+        status.get("Status")
+        for status in instance_status.get("InstanceStatus", {}).get("Details", [])
+        if status.get("Name") == "reachability"
+    )
```

Comment:
Does a reachability status always exist? And if so, does that mean we shouldn't need those defaults in the `.get` method calls? Otherwise we'll get a `StopIteration` thrown here if the comprehension creates a no-value generator:

```
>>> next(status.get("Status") for status in {}.get("InstanceStatus", {}).get("Details", []) if status.get("Name") == "reachability")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration
```

Either `StopIteration` needs to be anticipated and caught, or the default values need to be safe, or we should pass `next` a default. For example, `.get("Details")` could default to:

```
.get("Details", {"Name": "reachability", "Status": "unknown"})
```

Or better yet, use `next`'s optional default argument:

```suggestion
    status = next(
        (
            status.get("Status")
            for status in instance_status.get("InstanceStatus", {}).get("Details", [])
            if status.get("Name") == "reachability"
        ),
        None,
    )
```